### ISSUE TITLE
Use cargo-check in cargo maker

### ISSUE THREAD

**Issue by fuine** [neomake/neomake#843]
Cargo just got `check` command merged, which is what we wanted. It's essentially like `-Zno-trans`, so it should work significantly faster for the syntax checks. I will open the PR changing the maker after nightly version has been released, so that people who use nightly compiler could get it running now. For the actual merge i think that we should wait until this feature hits stable, and it's supposed to be introduced in 1.15, so in approximately 7 weeks. For more information please refer to [this PR](https://github.com/rust-lang/cargo/pull/3296).

**Comment by hpurmann** [comment]
Would love to see this 👍  Is there a PR I can check out, @fuine?

**Comment by fuine** [comment]
@hpurmann sorry, I haven't got enough time lately. Currently i can see that there is an ongoing WIP PR to use the json output: #942. What's more is that the `check` command has not been introduced in the 1.15 version, but is currently in beta, so we need to wait 7 weeks for its stabilization. After #942 gets resolved then I will open said PR.

**Comment by hpurmann** [comment]
Ok great, looking forward to that. Thank you for your work!

**Comment by jonhoo** [comment]
We'll probably want to monitor https://github.com/rust-lang/cargo/issues/3431 for when we'll be able to also check tests. I know for me at least that's a very important feature.

**Comment by gyscos** [comment]
`cargo-check` is now in rust stable. Tests are still not compiled, though.

**Comment by fuine** [comment]
I'm going to take a closer look at this issue over the weekend, and hopefully open a new PR.

**Comment by fuine** [comment]
It looks like there are two things that are relevant to this issue:
* should neomake opt into using `cargo-check` without test support, or should it stick to the current manner of calling `cargo test`? `cargo-check` is significantly faster, but on the other hand some people might want to compile their tests as well.
* I wouldn't like to create conflicts with #1123/#942. @blueyed @JelteF can you provide a rough estimation of when that PR could possibly land? To be fair the only change would modify one line, but I don't want to cause more troubles, seeing as that PR consists of quite a lot of commits.

**Comment by JelteF** [comment]
With regards to your second point #924 is mergeble now as far as I can tell
and works great. The #1123 PR requires more work, especially rewriting of
the tests.

The best way for this issue to be fixed right now in my opinion is by
giving the user the choice between a cargo full and a cargo check maker.
This can simply be done by copying the cargo maker definition and changing
the command. The only thing I'm not sure about is which one to have as the
default. I'd vote for check as the speed really makes it much nicer.

**Comment by fuine** [comment]
So do you think that `cargo-check` should be opted into in the current makers and you would port that over to the json PRs, or do you want to first land the json support and only after that should we switch to `cargo-check`?

I would also like to know what does @jonhoo think about `cargo-check` being the default, as he was the main voice of concern regarding tests support (not that I am against checking tests, but I think that we also should use `cargo-check` by default, and users which need to see the test results could use the current checker).

**Comment by jonhoo** [comment]
I'm very torn. I would strongly prefer moving to `cargo check` for the speed benefits (after all, this is why I currently use `timeout 2s cargo test --no-run` as my command), but giving up having tests checked seems unfortunate.

**Comment by jonhoo** [comment]
Actually, seems https://github.com/rust-lang/cargo/pull/3731 has landed, so `cargo check --all` should be available now. I think that's what we should use.

**Comment by jonhoo** [comment]
Eeeexcept apparently `--all` doesn't check tests yet (https://github.com/rust-lang/cargo/issues/3431#issuecomment-280708927) :(

I think my vote is for using `cargo check --all`, and then we'll eventually get checking of tests as well when that lands. Users like me who care deeply about the tests can keep the old command for now.

**Comment by fuine** [comment]
Please keep in mind that the `--all` flag is currently in beta, so we would need to wait for another 6 weeks to use that flag.

**Comment by jonhoo** [comment]
@fuine sure. I vote `cargo check` now, then `cargo check --all` in six weeks.

**Comment by JelteF** [comment]
#1123 has been merged, so now a pull request can be created to change the command to cargo check and maybe have a second maker for cargo build.

**Comment by blueyed** [comment]
> maybe have a second maker for cargo build

Or document how to change/configure the `args`.

**Comment by fuine** [comment]
I like the documentation idea, but you should remember that in order for the current `rustc`/`cargo` makers to work they need the `--message-format=json` argument to be in `args`. Can we ask users to always remember about that argument, or is there any clever trick to hide that argument away from the user that I am not seeing?

**Comment by blueyed** [comment]
@fuine 
It would be possible for the maker to have a `fn` property, which gets used as constructor, and could generate the `args`, i.e. add `--message-format=json` there always.
But when documenting it, it could just be included there - in the end it's the same: you either have to lookup and copy'n'paste a settings name to configure this behavior, or define it altogether.  I see that there are several other args though..

> 'args': ['test', '--no-run', '--message-format=json', '--quiet'],

So if 'test' can be replaced with `check` or `build` there, then a setting like `g:neomake_rust_cargo_method` could be used directly, or otherwise change the `args` accordingly.

Related: https://github.com/neomake/neomake/pull/1003.

**Comment by dpc** [comment]
I did `:PlugUpdate` and now I get:

```
Neomake: [3.-:-] project maker not found: cargo
```

My config:

```
autocmd! BufWritePost,BufEnter *.rs Neomake! cargo
```

Any ideas?

**Comment by JelteF** [comment]
Maybe just a more descriptive error, that the cargo maker is now the
default rust file maker. So that the regular rule to run neomake on save is
now enough.

**Comment by dpc** [comment]
That works. But `:cc`, `:cn` etc. don't.

**Comment by dpc** [comment]
Actually... it works for warnings, but not for errors...

**Comment by blueyed** [comment]
@dpc 
Sound like you should for a new issue?!

**Comment by fuine** [comment]
> >  'args': ['test', '--no-run', '--message-format=json', '--quiet'],
> 
> So if 'test' can be replaced with check or build there, then a setting like g:neomake_rust_cargo_method could be used directly, or otherwise change the args accordingly.
> 
> Related: #1003.

@blueyed 
The problem here is that for the `test` maker we'd want to use `test --no-run` arguments, whereas for the `check` maker we should probably settle on the single `check` (obviously both options should use the `--message-format=json --quiet` flags).  Please note that with the `1.17` version of Rust (scheduled 27.04) we should get the `--all` flag for `cargo-check`, which sadly does not support tests just yet.

To sum it up, this is how two argument arrays should look like for `test` and `check` respectively:
* `'args': ['test', '--no-run', '--message-format=json', '--quiet'],`
* `'args': ['check', '--message-format=json', '--quiet'],`

**Comment by blueyed** [comment]
@fuine 
Sounds like a config var for the maker makes sense then, e.g. `neomake_rust_cargo_mode`?!
It should use `neomake#GetSetting` (or wait for #1156), so that it can be configured on several levels.

**Comment by jonhoo** [comment]
I actually have the same issue as @dpc with quick-errors no longer working after this change..

**Comment by blueyed** [comment]
@jonhoo 
You mean after #1123?
You commented there already, but this is different?!
Can you provide steps to reproduce it?

**Comment by blueyed** [comment]
@dpc @jonhoo 
Location lists are used now.

**Comment by fuine** [comment]
@blueyed I'm not sure how to use `GetSetting` in the maker, as I couldn't find any example of its usage outside of `neomake.vim`. I wrote a working prototype using a simple 
`let maker_command = get(g:, 'neomake_rust_cargo_command', ['check'])`, because similar approach is used in the `javac` maker. Do you want me to open a PR with this approach or would you give me some pointers in here, as I don't know if this is acceptable?

**Comment by blueyed** [comment]
@fuine 
Please create a PR. It is fine like that, but should be a string, and not a list probably?!
You could also look at b: additionally.

**Comment by jonhoo** [comment]
@blueyed ahh, it's the switch from quickfix to location lists that tripped me up. thanks!

**Comment by fuine** [comment]
The PR for this issue has been merge. At the moment `cargo` maker will by default invoke `cargo-check` command. You can configure the exact command, along with its args via the `g:neomake_rust_cargo_command` and `b:neomake_rust_cargo_command` lists (e.g. to check everything, including test modules, put `let g:neomake_rust_cargo_command = ['test', '--no-run']` in your .vimrc file).

**Comment by jonhoo** [comment]
@fuine we may want to change the default to `cargo check --tests` now that rust-lang/cargo#3431 has seen some progress.

**Comment by fuine** [comment]
@jonhoo it looks like this is not supported in the stable release and as such I don't want to push things too rapidly. I think that [this PR](https://github.com/rust-lang/cargo/pull/4039) is what we want in the end, but until it lands in the stable I don't think neomake should change its current behaviour. Please note that you can locally set `g:neomake_rust_cargo_command` to what you prefer.

**Comment by blueyed** [comment]
Can we switch to using `cargo check --tests` by now?
(via https://github.com/neomake/neomake/issues/847#issuecomment-366076436)

**Comment by jonhoo** [comment]
I think this should really be `cargo check --all --all-targets`.

**Comment by fuine** [comment]
Agreed with @jonhoo, that looks like sensible default. Do you want me to open a PR for this?

**Comment by blueyed** [comment]
@fuine 
Yes, please.
